core/unit_test/TestGraph.hpp

```diff
@@ -840,4 +872,54 @@ TEST(TEST_CATEGORY, when_all_type) {
   static_assert(std::is_same_v<decltype(tail), node_ref_tail_t>);
 }
 
+TEST(TEST_CATEGORY, graph_then) {
+  using types = GraphNodeTypes<TEST_EXECSPACE>;
+
+  using view_t   = Kokkos::View<int, TEST_EXECSPACE>;
+  using memset_t = SetViewToValueFunctor<TEST_EXECSPACE, int>;
+  using then_t   = ThenFunctor<view_t>;
+  using policy_t =
+      Kokkos::RangePolicy<TEST_EXECSPACE, Kokkos::Impl::IsGraphKernelTag>;
+
+  using node_memset_t =
+      Kokkos::Impl::GraphNodeKernelImpl<TEST_EXECSPACE, policy_t, memset_t,
+                                        Kokkos::ParallelForTag>;
+  using node_ref_memset_t =
+      Kokkos::Experimental::GraphNodeRef<TEST_EXECSPACE, node_memset_t,
+                                         typename types::node_ref_root_t>;
+  using node_then_t = Kokkos::Impl::GraphNodeThenImpl<TEST_EXECSPACE, then_t>;
+  using node_ref_then_t =
+      Kokkos::Experimental::GraphNodeRef<TEST_EXECSPACE, node_then_t,
+                                         node_ref_memset_t>;
```

Comment:
That's a lot of churn for checking the type of the "then" node below.
Why is it important to have it?

Reply:
I think it's worth that at least once, we check the full type of the nodes. This ensures that the type erasure is exactly where we want it.